src/containers/ap/ApGuestRegister.tsx

```diff
@@ -0,0 +1,316 @@
+import * as React from "react";
+import { setAPImage } from "../../util/set-bg-image";
+import FactaArticleRegion from "../../theme/facta/FactaArticleRegion";
+import { History } from 'history'
+import DateTriPicker, {  DateTriPickerProps } from "@components/DateTriPicker";
+import PhoneTriBox, { PhoneTriBoxProps } from "@components/PhoneTriBox";
+import formUpdateState from '../../util/form-update-state';
+import TextInput from '../../components/TextInput';
+import Validation from '../../util/Validation';
+import FactaButton from '../../theme/facta/FactaButton';
+import { Option, none } from 'fp-ts/lib/Option';
+//import { jpBasePath } from "../../app/paths/jp/_base";
+//import { Link } from "react-router-dom";
+//import { apCreateAcctRoute } from "../../app/routes/ap/create-acct";
+import {FactaErrorDiv} from "@facta/FactaErrorDiv";
+import FactaMainPage from "../../theme/facta/FactaMainPage";
+import * as moment from 'moment';
+import range from "@util/range";
+
+type Props = {
+	history: History<any>
+}
+
+
+type State = {
+	formData: Form,
+	validationErrors: string[],
+	pageState: PageState
+}
+
+enum PageState{
+	GUEST_INFORMATION,
+	EMERGENCY_CONTACT,
+	WAIVER,
+	FINISH
+}
+
+const defaultForm = {
+	firstName: none as Option<string>,
+	lastName: none as Option<string>,
+	email: none as Option<string>,
+	dobMonth: none as Option<string>,
+	dobDay: none as Option<string>,
+	dobYear: none as Option<string>,
+	guestPhoneFirst: none as Option<string>,
+	guestPhoneSecond: none as Option<string>,
+	guestPhoneThird: none as Option<string>,
+	guestPhoneExt: none as Option<string>,
+	guestPhoneType: none as Option<string>,
+	ecFirstName: none as Option<string>,
+	ecLastName: none as Option<string>,
+	ecPhoneFirst: none as Option<string>,
+	ecPhoneSecond: none as Option<string>,
+	ecPhoneThird: none as Option<string>,
+	ecPhoneExt: none as Option<string>,
+	ecPhoneType: none as Option<string>,
+	ecRelationship: none as Option<string>
+
+}
+
+type Form = typeof defaultForm
+
+class FormInput extends TextInput<Form> {}
+
+const validate: (state: State) => string[] = state => {
+	const isNotNull = (os: Option<string>) => os.isSome() && os.getOrElse("").length > 0;
+	//const emailValid = (os : Option<string>) => os.isSome() && os.getOrElse("")
+	const validations = [
```

Comment:
My gut reaction is to say, don't bother doing too many validations client-side, most of these things probably really belong server-side (otherwise someone could always just curl the create-person endpoint directly and bypass all this logic in the client).  And once the logic lives on the server, it doesn't also need to live on the client, hence why there aren't a huge number of validations on other pages.  This is not an issue for a closed client like the kiosk where the endpoints are authenticated, but for a web client it is

That said, actually doing these validations server side would be awkward, since I'm assuming you guys won't actually call the server until the final submit on the last page.  We could do a sort of "preflight" api call when they finish the info screen, and have the server save the fields as a proto-person record, and then when they submit the final screen it turns that record into a real person.  But that all seems like overkill

So I think that is all to say, doing these validations on the client and not on the server opens us up to that risk of someone posting to the api server directly without using the client and skipping the validations, but in this case I think the cost of doing it right outweighs the benefit, so let's leave this as is.

Validations that the server runs on the member page which you should replicate here are

- DOB is a valid date (e.g. 2/31/1990 should throw an error)
- Email is a valid email (I use the regex  /^[A-Z0-9._%-]+@[A-Z0-9._%-]+\.[A-Z]{2,4}$/i  )
- Phone number is valid (matches  /^[0-9]{10}.*$/  ie exactly 10 digits, with the .* for any extension )

https://github.com/community-boating/cbidb-api/blob/master/app/org/sailcbi/APIServer/Api/Endpoints/Member/APRequiredInfo.scala